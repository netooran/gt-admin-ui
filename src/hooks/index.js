export { default as useDarkMode } from './useDarkMode/useDarkMode';
