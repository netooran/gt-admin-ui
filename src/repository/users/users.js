import createResource from 'utils/resource/resource';

const fetchUsers = async () => {
  const response = await fetch(
    'https://geektrust.s3-ap-southeast-1.amazonaws.com/adminui-problem/members.json'
  );
  return await response.json();
};

export const fetchUsersResource = () => createResource(fetchUsers());
