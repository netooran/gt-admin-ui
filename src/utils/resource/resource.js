const createResource = (promise) => {
  let status = 'loading';
  let result;

  const suspender = promise.then(
    (data) => {
      status = 'success';
      result = data;
    },
    (error) => {
      status = 'error';
      result = error;
    }
  );

  return {
    read() {
      if (status === 'loading') throw suspender;
      else if (status === 'error') throw result;
      else return result;
    },
  };
};

export const createMockResource = (data, error) => ({
  loading: {
    read: () => {
      throw Promise.resolve();
    },
  },
  error: {
    read: () => {
      throw new Error(error);
    },
  },
  success: {
    read: () => data,
  },
});

export default createResource;
