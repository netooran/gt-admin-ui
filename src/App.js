import { Suspense } from 'react';
import styled from 'styled-components';
import Spinner from 'shared/Spinner/Spinner';
import Table from 'shared/Table/Table';
import { ViewPort, Center } from 'shared/Containers/Containers';
import validators, { getValidator } from './utils/validators/validators';
import { fetchUsersResource } from 'repository/users/users';
import ErrorBoundary from 'shared/ErrorBoundary/ErrorBoundary';

const TableWrapper = styled.div`
  height: 50vw;
  min-width: 1000px;
`;

const App = () => {
  const userResource = fetchUsersResource();

  const columns = [
    {
      name: 'name',
      type: 'text',
      validate: getValidator([validators.isNotEmpty], 'name'),
    },
    {
      name: 'email',
      type: 'email',
      validate: getValidator([validators.isValidEmail], 'email'),
    },
    {
      name: 'role',
      type: 'text',
      validate: getValidator([validators.isValidUserRole], 'role'),
    },
  ];

  const options = {
    editable: true,
    deletable: true,
    selectable: true,
    paginatable: true,
    searchable: true,
  };

  return (
    <ViewPort>
      <Center>
        <TableWrapper>
          <ErrorBoundary>
            <Suspense fallback={<Spinner />}>
              <Table
                aria-label="Users table"
                dataResource={userResource}
                columns={columns}
                options={options}
              />
            </Suspense>
          </ErrorBoundary>
        </TableWrapper>
      </Center>
    </ViewPort>
  );
};

export default App;
