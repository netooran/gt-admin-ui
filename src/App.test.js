import { render } from '@testing-library/react';
import App from './App';
import { createMockResource } from 'utils/resource/resource';
import { fetchUsersResource } from 'repository/users/users';
import { useDarkMode } from 'hooks';

jest.mock('repository/users/users');
jest.mock('hooks');

useDarkMode.mockReturnValue([true, jest.fn()]);

const mockResource = createMockResource(
  [{ id: '1', name: 'Alex', email: 'alex@protonmail.com', role: 'admin' }],
  'Api error'
);

describe.skip('App', () => {
  test('renders loader', () => {
    fetchUsersResource.mockReturnValue(mockResource.loading);
    const { container } = render(<App />);
    expect(container).toMatchSnapshot();
  });

  test('renders table when data is available', () => {
    fetchUsersResource.mockReturnValue(mockResource.success);

    const { container } = render(<App />);
    expect(container).toMatchSnapshot();
  });

  test('renders error message on api failure', () => {
    fetchUsersResource.mockReturnValue(mockResource.error);

    const { container } = render(<App />);
    expect(container).toMatchSnapshot();
  });
});
