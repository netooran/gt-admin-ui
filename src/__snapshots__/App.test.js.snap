// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders error message on api failure 1`] = `
.c4 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c0 {
  width: 100vw;
  height: 100vh;
}

.c1 {
  display: grid;
  place-items: center;
  width: 100%;
  height: 100%;
}

.c3 {
  width: 100%;
  height: 100%;
}

.c3 div {
  text-align: center;
}

.c3 h1 {
  color: darkgray;
}

.c3 svg {
  fill: #e5e5e5;
}

.c2 {
  height: 50vw;
  min-width: 1000px;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        <div
          class="c1 c3"
        >
          <div>
            <svg
              aria-hidden="true"
              class="c4"
              fill="currentColor"
              focusable="false"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M15.165 15.582c4.587-4.073 8.141-9.424 6.057-11.771-.661-.744-1.584-1.089-2.575-.983-.832.091-1.687.502-2.549 1.192a8.922 8.922 0 00-8.712.282 8.917 8.917 0 00-4.109 5.515 8.892 8.892 0 00.306 5.325c-1.065 1.203-2.054 3.677-.823 5.063.517.581 1.257.841 2.147.841 2.707 0 6.808-2.399 10.258-5.464zm3.699-10.767c.358-.034.632.064.861.323.231.261.169.946-.252 1.929a9.059 9.059 0 00-1.617-1.853c.431-.262.776-.373 1.008-.399zM4.633 17.118a8.979 8.979 0 001.568 1.737c-1.025.303-1.714.283-1.945.021-.217-.243.002-1.069.377-1.758zm16.31-5.869c-1.215 1.797-2.906 3.671-4.778 5.333-1.934 1.719-4.066 3.208-6.05 4.202a9.082 9.082 0 001.874.212 8.986 8.986 0 004.616-1.282 8.915 8.915 0 004.338-8.465z"
              />
            </svg>
            <h1>
              Something went wrong
            </h1>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders loader 1`] = `
.c4 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c0 {
  width: 100vw;
  height: 100vh;
}

.c1 {
  display: grid;
  place-items: center;
  width: 100%;
  height: 100%;
}

.c3 {
  -webkit-animation: spin 2s linear infinite;
  animation: spin 2s linear infinite;
}

.c3 svg {
  width: 3rem;
}

.c2 {
  height: 50vw;
  min-width: 1000px;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        <div
          aria-label="Loading"
          class="c1 c3"
        >
          <svg
            aria-hidden="true"
            class="c4"
            fill="currentColor"
            focusable="false"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M2 11h5v2H2zm15 0h5v2h-5zm-6 6h2v5h-2zm0-15h2v5h-2zM4.222 5.636l1.414-1.414 3.536 3.536-1.414 1.414zM19.778 18.364l-1.414 1.414-3.536-3.536 1.414-1.414zM7.758 14.828l1.414 1.414-3.536 3.536-1.414-1.414zM14.828 7.757l3.536-3.535 1.414 1.415-3.536 3.535z"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders table when data is available 1`] = `
.c13 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c0 {
  width: 100vw;
  height: 100vh;
}

.c1 {
  display: grid;
  place-items: center;
  width: 100%;
  height: 100%;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c5 {
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c6 {
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  border: 1px solid #d8d8d8;
  border-radius: 0.2rem;
  background-color: white;
  min-width: 250px;
  padding: .25rem 1rem;
}

.c6:focus-within {
  border: 1px solid #1890ff;
}

.c6 svg {
  fill: red;
  width: 1.2rem;
  padding: 0 0.5rem;
}

.c10 {
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  border: 1px solid transparent;
  border-radius: 0.2rem;
  background-color: white;
  min-width: 250px;
  background-color: transparent;
  pointer-events: none;
}

.c10:focus-within {
  border: 1px solid #1890ff;
}

.c10 svg {
  fill: red;
  width: 1.2rem;
  padding: 0 0.5rem;
}

.c7 {
  border-style: none;
  border-radius: 0.2rem;
  padding: 0.5rem;
  width: 100%;
}

.c7:focus {
  outline: none;
}

.c7:disabled {
  color: inherit;
  background-color: transparent;
  pointer-events: none;
}

.c3 {
  padding: 1rem;
}

.c12 {
  --button-primary-color: #000000;
  border: none;
  cursor: pointer;
  opacity: 0.7;
  font-weight: bold;
  padding: 0.5rem 1.5rem;
  border-radius: 3rem;
  background-color: transparent;
  color: var(--button-primary-color);
  padding: 0;
}

.c12 svg {
  width: 1.2rem;
}

.c12:disabled {
  opacity: 0.4;
  pointer-events: none;
}

.c12:hover {
  opacity: 1;
}

.c14 {
  --button-primary-color: #000000;
  border: none;
  cursor: pointer;
  opacity: 0.7;
  font-weight: bold;
  padding: 0.5rem 1.5rem;
  border-radius: 3rem;
  --button-primary-color: red;
  background-color: transparent;
  color: var(--button-primary-color);
  padding: 0;
}

.c14 svg {
  width: 1.2rem;
}

.c14:disabled {
  opacity: 0.4;
  pointer-events: none;
}

.c14:hover {
  opacity: 1;
}

.c16 {
  --button-primary-color: #000000;
  border: none;
  cursor: pointer;
  opacity: 0.7;
  font-weight: bold;
  padding: 0.5rem 1.5rem;
  border-radius: 3rem;
  --button-primary-color: red;
  background-color: var(--button-primary-color);
  color: white;
}

.c16 svg {
  width: 1.2rem;
}

.c16:disabled {
  opacity: 0.4;
  pointer-events: none;
}

.c16:hover {
  opacity: 1;
}

.c17 {
  --button-primary-color: #000000;
  border: none;
  cursor: pointer;
  opacity: 0.7;
  font-weight: bold;
  padding: 0.5rem 1.5rem;
  border-radius: 3rem;
  background-color: var(--button-primary-color);
  color: white;
  min-width: 3rem;
  min-height: 3rem;
  padding: 0;
}

.c17 svg {
  width: 1.2rem;
}

.c17:disabled {
  opacity: 0.4;
  pointer-events: none;
}

.c17:hover {
  opacity: 1;
}

.c19 {
  --button-primary-color: #000000;
  border: none;
  cursor: pointer;
  opacity: 0.7;
  font-weight: bold;
  padding: 0.5rem 1.5rem;
  border-radius: 3rem;
  background-color: transparent;
  border: 1px solid var(--button-primary-color);
  color: var(--button-primary-color);
  min-width: 3rem;
  min-height: 3rem;
  padding: 0;
}

.c19 svg {
  width: 1.2rem;
}

.c19:disabled {
  opacity: 0.4;
  pointer-events: none;
}

.c19:hover {
  opacity: 1;
}

.c11 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c11 button {
  margin-right: 0.5rem;
}

.c18 {
  --button-primary-color: #1890ff;
}

.c15 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 1rem;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c9 {
  --row-active-bg: #f2f2f2;
}

.c9:hover {
  background-color: var(--row-active-bg);
}

.c8 {
  text-align: left;
  border-collapse: collapse;
  width: 100%;
}

.c8 th {
  text-transform: capitalize;
}

.c8 tr {
  border-bottom: 1px solid lightgray;
}

.c8 td,
.c8 th {
  padding: 0.75rem 1rem;
}

.c2 {
  height: 50vw;
  min-width: 1000px;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        <div>
          <div
            class="c3"
          >
            <div
              aria-label="Search"
              class="c4 c5 c6"
              placeholder="Search by name, email, role"
            >
              <input
                aria-label="Search"
                class="c7"
                placeholder="Search by name, email, role"
                role="search"
                type="text"
                value=""
              />
            </div>
          </div>
          <table
            aria-label="Users table"
            class="c8"
          >
            <thead>
              <tr>
                <th>
                  <input
                    aria-label="Select all"
                    type="checkbox"
                  />
                </th>
                <th>
                  name
                </th>
                <th>
                  email
                </th>
                <th>
                  role
                </th>
                <th>
                  actions
                </th>
              </tr>
            </thead>
            <tbody>
              <tr
                class="c9"
              >
                <td>
                  <input
                    aria-label="Select row"
                    type="checkbox"
                  />
                </td>
                <td>
                  <div
                    aria-label="name"
                    class="c4 c5 c10"
                  >
                    <input
                      aria-label="name"
                      class="c7"
                      disabled=""
                      role="textbox"
                      type="text"
                      value="Alex"
                    />
                  </div>
                </td>
                <td>
                  <div
                    aria-label="email"
                    class="c4 c5 c10"
                  >
                    <input
                      aria-label="email"
                      class="c7"
                      disabled=""
                      role="textbox"
                      type="email"
                      value="alex@protonmail.com"
                    />
                  </div>
                </td>
                <td>
                  <div
                    aria-label="role"
                    class="c4 c5 c10"
                  >
                    <input
                      aria-label="role"
                      class="c7"
                      disabled=""
                      role="textbox"
                      type="text"
                      value="admin"
                    />
                  </div>
                </td>
                <td>
                  <div
                    class="c11"
                  >
                    <button
                      aria-label="Edit"
                      class="c12"
                    >
                      <svg
                        aria-hidden="true"
                        class="c13"
                        fill="currentColor"
                        focusable="false"
                        viewBox="0 0 24 24"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M18.988 2.012l3 3L19.701 7.3l-3-3zM8 16h3l7.287-7.287-3-3L8 13z"
                        />
                        <path
                          d="M19 19H8.158c-.026 0-.053.01-.079.01-.033 0-.066-.009-.1-.01H5V5h6.847l2-2H5c-1.103 0-2 .896-2 2v14c0 1.104.897 2 2 2h14a2 2 0 002-2v-8.668l-2 2V19z"
                        />
                      </svg>
                    </button>
                    <button
                      aria-label="Delete"
                      class="c14"
                    >
                      <svg
                        aria-hidden="true"
                        class="c13"
                        fill="currentColor"
                        focusable="false"
                        viewBox="0 0 24 24"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M6 7H5v13a2 2 0 002 2h10a2 2 0 002-2V7H6zm4 12H8v-9h2v9zm6 0h-2v-9h2v9zm.618-15L15 2H9L7.382 4H3v2h18V4z"
                        />
                      </svg>
                    </button>
                  </div>
                </td>
              </tr>
            </tbody>
          </table>
          <div
            class="c15"
          >
            <button
              aria-label="Delete selected"
              class="c16"
              disabled=""
            >
              Delete Selected
            </button>
            <div
              class="c11"
            >
              <button
                aria-label="Go to first page"
                class="c17 c18"
                disabled=""
                shape="round"
              >
                <svg
                  aria-hidden="true"
                  class="c13"
                  fill="currentColor"
                  focusable="false"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M8.121 12l4.94-4.939-2.122-2.122L3.879 12l7.06 7.061 2.122-2.122z"
                  />
                  <path
                    d="M17.939 4.939L10.879 12l7.06 7.061 2.122-2.122L15.121 12l4.94-4.939z"
                  />
                </svg>
              </button>
              <button
                aria-label="Go to previous page"
                class="c17 c18"
                disabled=""
                shape="round"
              >
                <svg
                  aria-hidden="true"
                  class="c13"
                  fill="currentColor"
                  focusable="false"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M13.939 4.939L6.879 12l7.06 7.061 2.122-2.122L11.121 12l4.94-4.939z"
                  />
                </svg>
              </button>
              <button
                aria-label="Go to page 1"
                class="c19 c18"
                shape="round"
              >
                1
              </button>
              <button
                aria-label="Go to next page"
                class="c17 c18"
                disabled=""
                shape="round"
              >
                <svg
                  aria-hidden="true"
                  class="c13"
                  fill="currentColor"
                  focusable="false"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M10.061 19.061L17.121 12l-7.06-7.061-2.122 2.122L12.879 12l-4.94 4.939z"
                  />
                </svg>
              </button>
              <button
                aria-label="Go to last page"
                class="c17 c18"
                disabled=""
                shape="round"
              >
                <svg
                  aria-hidden="true"
                  class="c13"
                  fill="currentColor"
                  focusable="false"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M13.061 4.939l-2.122 2.122L15.879 12l-4.94 4.939 2.122 2.122L20.121 12z"
                  />
                  <path
                    d="M6.061 19.061L13.121 12l-7.06-7.061-2.122 2.122L8.879 12l-4.94 4.939z"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
